django/db/models/expressions.py

```diff
@@ -15,6 +15,25 @@
 from django.utils.hashable import make_hashable
 
 
+class MySQLDateTimeMixin:
+    """
+    Some expressions with output_field set to DateField(), DateTimeField(), or
+    TimeField() must be cast to the appropriate database type to avoid the
+    string value passed to the database being returned as a string instead of
+    the expected type.
+    """
+    def as_mysql(self, compiler, connection):
+        sql, params = self.as_sql(compiler, connection)
+        field_names = ['DateField', 'DateTimeField', 'TimeField']
+        try:
+            if self.output_field.get_internal_type() in field_names:
+                db_type = self.output_field.cast_db_type(connection)
+                sql = 'CAST(%s AS %s)' % (sql, db_type)
+        except FieldError:
```

Comment:
Which line(s) above can raise `FieldError`?

Reply:
Good question. I didn't check, but this follows the exact same pattern as `SQLiteNumericMixin`. Will check if it's needed…

Reply:
So apparently it is possible for `FieldError` to be raised when accessing `.output_field`. See 3bbaf84d6533fb61ac0038f2bbe52ee0d7b4fd10 and the review in https://github.com/django/django/pull/4828#discussion_r32477199. `Value` inherits from `Expression` which inherits from `BaseExpression`. The property `BaseExpression.output_field` can raise `FieldError`.

Reply:
Okay, then I would suggest putting the try-except around `output_field = self.output_field` if that's what it's protecting. Also, I think you can do away with the separate `field_names` variable.

Reply:
Have made the change. (I think it makes it more clunky and less consistent with `SQLiteNumericMixin`, but 🤷🏻‍♂️.)

Reply:
I think it would look less clunky to return early in the `except` clause. Then the `else` isn't needed and you can reduce the indent in the remainder of the function. I think it does help the reader to scope `except`'s as narrowly as possible, because then the purpose is clear, and it avoids swallowing exceptions you might not have known about.